### PR TITLE
Find and fix vulnerable npm packages

### PR DESCRIPTION
**Pull request by timtebeek**
## What's changed?
Add a recipe to Find and fix vulnerable npm packages

## Anything in particular you'd like reviewers to focus on?
1. Do the test input and output align with what we hope to see? `^` versions specifically.
2. Search markers are not yet implemented; do we need them for v1? That would show `.minor` and `.major` CVEs in code that now only show up in data table rows
3. Like `DependencyInsight` this for now assumes a single `package-lock.json`. Should we support multiple?

## Anyone you would like to review specifically?
@zieka 

## Have you considered any alternatives or workarounds?
- This copies elements from rewrite-java-dependencies, in particular how we parse the GitHub advisories database. In the copy here we parse for the NPM ecosystem. Seemed odd to parse these in `rewrite-java-dependencies`, unless we rename that `rewrite-dependencies`.
- Significant changes as to the `DependencyVulnerabilityCheck` for Java, producing different data table rows sans groupId/artifactId split, and without dependency depth, due to lacking data. This will require new visualizations, and separate recipe runs, but seemed a stretch to "make it fit" the existing model in rewrite-java-dependencies.

### CODE REVIEW

File: .github/workflows/advisories.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          title: "[Auto] GitHub advisories as of ${{ env.NOW }}"
+          body: |
+            [Auto] GitHub advisories as of ${{ env.NOW }}.
+          commit-message: "[Auto] GitHub advisories as of ${{ env.NOW }}"
+          labels: enhancement
+      - name: Check outputs
+        if: ${{ steps.cpr.outputs.pull-request-number }}
+        run: |
+          echo "Pull Request Number - ${{ steps.cpr.outputs.pull-request-number }}"
+          echo "Pull Request URL - ${{ steps.cpr.outputs.pull-request-url }}"
```
Comment (timtebeek) [inline .github/workflows/advisories.yml:56]: We might just want to skip the PRs here and go for a direct commit, as we have seen work well on the maven plugin:
https://github.com/openrewrite/rewrite-maven-plugin/blob/db055e402890e82385b897e1faec1a7f6e5c7786/.github/workflows/bump-snapshots.yml#L32

**Comment by timtebeek** [comment]
If we want to reduce the duplication, I've now made the ecosystem of the GitHub security advisory database configurable in
- https://github.com/openrewrite/rewrite-java-dependencies/pull/112

We could then push that file here from that repository, as we also do with a few other deploy keys such as for our docs.

File: src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        parseAdvisories(advisoriesRepo, advisoriesCsv);
+    }
+
+    static void parseAdvisories(File advisoriesRepoInput, File advisoriesCsvOutput) throws IOException {
+        try (FileOutputStream fos = new FileOutputStream(advisoriesCsvOutput)) {
+            Files.walkFileTree(advisoriesRepoInput.toPath(), emptySet(), 16, new MavenAdvisoriesVisitor(fos));
+        }
+    }
+
+    private static final class MavenAdvisoriesVisitor extends SimpleFileVisitor<Path> {
```
Comment (shanman190) [inline src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java:66]:
```suggestion
    private static final class NpmAdvisoriesVisitor extends SimpleFileVisitor<Path> {
```

Thoughts?

File: src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            System.err.println("Advisories CSV " + advisoriesCsv + " not writable");
+            System.exit(1);
+        }
+
+        parseAdvisories(advisoriesRepo, advisoriesCsv);
+    }
+
+    static void parseAdvisories(File advisoriesRepoInput, File advisoriesCsvOutput) throws IOException {
+        try (FileOutputStream fos = new FileOutputStream(advisoriesCsvOutput)) {
+            Files.walkFileTree(advisoriesRepoInput.toPath(), emptySet(), 16, new MavenAdvisoriesVisitor(fos));
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java:62]:
```suggestion
            Files.walkFileTree(advisoriesRepoInput.toPath(), emptySet(), 16, new NpmAdvisoriesVisitor(fos));
```

File: src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            Files.walkFileTree(advisoriesRepoInput.toPath(), emptySet(), 16, new NpmAdvisoriesVisitor(fos));
+        }
+    }
+
+    private static final class NpmAdvisoriesVisitor extends SimpleFileVisitor<Path> {
+        private final FileOutputStream fos;
+        private final ObjectMapper reader;
+        private final ObjectWriter writer;
+
+        public MavenAdvisoriesVisitor(FileOutputStream fos) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/nodejs/github/ParseAdvisories.java:71]:
```suggestion
        public NpmAdvisoriesVisitor(FileOutputStream fos) {
```

File: src/main/java/org/openrewrite/nodejs/github/Vulnerability.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.time.ZonedDateTime;
+
+@Value
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+@JsonPropertyOrder({"cve", "published", "summary", "packageName", "introducedVersion", "fixedVersion", "severity", "cwes"})
+public class Vulnerability {
+    @EqualsAndHashCode.Include
+    String cve;
+
+    ZonedDateTime published;
```
Comment (pstreef) [inline src/main/java/org/openrewrite/nodejs/github/Vulnerability.java:31]: If possible (and not breaking an existing standard) I would rename this to `publishedAt` so from the name you expect a timestamp (and not maybe a boolean).

File: src/main/java/org/openrewrite/nodejs/github/advisories/Advisory.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+@Value
+public class Advisory {
+    String id;
+    ZonedDateTime published;
+    List<Affected> affected;
+    List<String> aliases;
+    String summary;
+
+    @JsonProperty("database_specific")
```
Comment (pstreef) [inline src/main/java/org/openrewrite/nodejs/github/advisories/Advisory.java:32]: maybe just set the objectMapper to accept snake case? Or set this on the class level.

File: src/main/java/org/openrewrite/nodejs/github/advisories/Range.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import lombok.Value;
+import org.openrewrite.internal.lang.Nullable;
+
+import java.util.List;
+
+@Value
+public class Range {
+    List<Event> events;
+
+    public @Nullable String getIntroduced() {
```
Comment (pstreef) [inline src/main/java/org/openrewrite/nodejs/github/advisories/Range.java:27]: ### ```suggestion
    @Nullable
    public String getIntroduced() {
```

File: src/main/java/org/openrewrite/nodejs/github/advisories/Range.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        for (Event event : events) {
+            if (event.getIntroduced() != null) {
+                return event.getIntroduced();
+            }
+        }
+        return null;
+    }
+
+    public @Nullable String getFixed() {
```
Comment (pstreef) [inline src/main/java/org/openrewrite/nodejs/github/advisories/Range.java:36]:
```suggestion
    @Nullable
    public String getFixed() {
```

**Review by pstreef — APPROVED** [review]
Approved with some minor comments.
edit: Oh I guess this is still in draft. Feel free to re-request 😅

File: src/main/java/org/openrewrite/nodejs/github/advisories/Range.java (hunk 7 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/nodejs/github/advisories/Range.java:27]: That's not [how the cool kids do it](https://docs.openrewrite.org/recipes/staticanalysis/nullableonmethodreturntype)! 🙃

File: src/main/java/org/openrewrite/nodejs/github/Vulnerability.java (hunk 5 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/nodejs/github/Vulnerability.java:31]: Applied in https://github.com/openrewrite/rewrite-java-dependencies/pull/112, thanks!

File: src/main/java/org/openrewrite/nodejs/github/advisories/Advisory.java (hunk 6 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/nodejs/github/advisories/Advisory.java:32]: Applied in https://github.com/openrewrite/rewrite-java-dependencies/pull/112, thanks!

**Pull request merged**